### DISCUSSION TITLE
How to delete patches and banks?

### DISCUSSION THREAD

**Discussion by RadekPilich** [christofmuc/KnobKraft-orm#378] [Q&A] [answered]
Trying to delete a whole imported bank I get the following error:

error DATABASE ERROR in deletePatches via filter: SQL Exception FOREIGN KEY constraint failed

------

Deleting patches also doesn't seem to work and I get the following error:

warning Could not delete patch Synthwave 2.3 from database as it is referred to be at least one bank definition. Removed it from user lists and set it to hidden instead

**Comment by christofmuc** [comment]
This must be a bug, and it is good that the database protects the integrity of the data. 

Can you share some more details, or maybe even the database itself? That'll be the fastest way to resolve this.

**Comment by RadekPilich** [comment]
Here is the db file. It's only a bunch of initial test imports from the compatible synths I have.

DELETED

**Reply by RadekPilich** [reply]
@christofmuc Have you noticed my upload? i'd like to delete the uploaded attachment ASAP, as there are some purchased soundpacks in there.

**Reply by christofmuc** [reply]
Ups, sorry, yes, downloaded it now, you can delete!

**Reply by christofmuc** [reply]
I will investigate - I made an issue #386 to track it.

**Reply by christofmuc** [reply, marked as answer]
Ok, I think I understand the problem: The patch synthwave 2.3 is part of the bank 4, which has been loaded from the synth. As it represents the synth's memory, it should not be deleted from the database, because then Knobkraft would have a hole in the synth's bank.

The workaround: You can make Knobkraft forget the sync with the synth, in this case by dragging the synth bank 4 onto the trash. This will not delete any patches yet, but remove the synced bank, it will turn red again as in "unknown" (you can resync the bank from the synth later). 

Then, nothing should stop you from e..g setting the filter to show only hidden patchs, and drag the synthwave 2,3 to the trash. Then it will be gone. 

I have to think about this, obviously throwing the database error is not totally helpful, but I wonder what the system should actually do. Maybe it is best if it just checks if the patch is currently in the synth, and refuse to delete it...

**Comment by christofmuc** [comment]
Closing this as answered, the bug itself is tracked at #386.
